### PR TITLE
Enable service preauthorization and authentication to work with profiles

### PR DESCRIPTION
**Pull request by shamsimam**
## Changes proposed in this PR

- moves profile computation to merge-defaults
- adds support for token pre-authorization with profiles providing permitted-user field
- curry discover-service-parameters in routines
- adds profiles support in service discovery
- includes x-waiter headers in discovery service-description-parameters

## Why are we making these changes?

Allows service pre-authorization with tokens that use profiles.

### CODE REVIEW

File: waiter/src/waiter/service_description.clj [hunk 1]
```diff
@@ -886,11 +888,12 @@
           (if loop-token
             (let [token-data (token->token-data kv-store loop-token token-data-keys true false)]
               (recur (assoc loop-token->token-data loop-token token-data) remaining-tokens))
-            (compute-service-description-template-from-tokens
-              attach-token-defaults-fn token-sequence loop-token->token-data))))
+            (let []
```
Comment (sradack) [inline waiter/src/waiter/service_description.clj:891]: Can we look at the diff before we open it for review?  Can we minimize unnecessary changes?  Moving functions, etc.

Comment (shamsimam) [reply waiter/src/waiter/service_description.clj:891]: Sorry about this `let`, will fix.
Moving `merge-defaults` and `default-and-override` was necessary. Any other you felt moved unnecessarily?

Comment (shamsimam) [reply waiter/src/waiter/service_description.clj]: Removed the empty `let` block.

File: waiter/src/waiter/core.clj [hunk 2]
```diff
@@ -600,21 +600,21 @@
    - (on-disabled) when the authentication is disabled
    - (on-auth-required) when authentication is enabled."
   [waiter-discovery on-error on-disabled on-auth-required]
-  (let [{:keys [service-parameter-template token waiter-headers]} waiter-discovery
-        {:strs [authentication] :as service-description} service-parameter-template
+  (let [{:keys [service-description-template token waiter-headers]} waiter-discovery
```
[the remark this answers is not in the document]
Comment (shamsimam) [reply waiter/src/waiter/core.clj:603]: `service-parameter-template` comes from token.
`service-description(-template)` comes from token and headers.

File: waiter/src/waiter/service_description.clj [hunk 3]
```diff
@@ -863,21 +863,23 @@
...
 (defn- prepare-service-description-template-from-tokens
   "Prepares the service description using the token(s)."
-  [waiter-headers request-headers kv-store waiter-hostnames attach-token-defaults-fn]
+  [waiter-headers request-headers kv-store waiter-hostnames attach-service-defaults-fn attach-token-defaults-fn]
   (let [{:keys [token source]}
         (retrieve-token-from-service-description-or-hostname waiter-headers request-headers waiter-hostnames)]
     (cond
       (= source :host-header)
-      (let [token-data (token->token-data kv-store token token-data-keys false false)]
+      (let [token-data (token->token-data kv-store token token-data-keys false false)
```
[the remark this answers is not in the document]
Comment (shamsimam) [reply waiter/src/waiter/service_description.clj:878]: Parameters to the invocation of `compute-service-description-template-from-tokens` (which  changed) that make it easy to view all three invocations in the `cond` branches.

File: waiter/src/waiter/service_description.clj [hunk 4]
```diff
@@ -975,27 +977,33 @@
             sanitized-service-description (apply dissoc service-description metadata-keys)]
         (assoc sanitized-service-description "metadata" renamed-metadata-map)))))
 
+(defn- waiter-headers->service-parameters
```
[the remark this answers is not in the document]
Comment (shamsimam) [reply waiter/src/waiter/service_description.clj:980]: Necessary, this function is invoked in two places now.

File: waiter/src/waiter/service_description.clj [hunk 5]
```diff
@@ -1331,17 +1339,30 @@
 (defn discover-service-parameters
```
[the remark this answers is not in the document]
Comment (shamsimam) [reply waiter/src/waiter/service_description.clj:1339]: Previously the defaults didn't include values for fields we were looking at in `service-parameter-template`, so excluding them didn't hurt. 
With profiles, the values can now come from the profile defaults and so we need to include the profile defaults.
Using the common `attach-service-defaults-fn` avoids needing to differentiate defaults and profile defaults.

It is incorrect (and imo a bug) to look at token parameters for values that could be overridden by request headers.
E.g. JWT is disabled by env variable in the req header, but we incorrectly assume by looking at the token definition that JWT is enabled or vice versa.
Including the fields from headers fixes that issue.

File: waiter/test/waiter/auth/composite_test.clj [hunk 6]
```diff
@@ -48,10 +48,8 @@
    :default-authentication-provider "one-user"})
 
 (defn dummy-composite-authenticator
-  ([config]
-   (composite-authenticator (assoc config :default-authentication "standard")))
-  ([]
-   (dummy-composite-authenticator valid-config)))
+  ([] (dummy-composite-authenticator valid-config))
```
[the remark this answers is not in the document]
Comment (shamsimam) [reply waiter/test/waiter/auth/composite_test.clj:51]: Sorry.

File: waiter/test/waiter/core_test.clj [hunk 7]
```diff
@@ -1131,9 +1139,16 @@
...
+    (kv/store kv-store "www.token-1s.com" {"cpu" 1
+                                           "mem" 2048
+                                           "profile" "service"})
     (kv/store kv-store "www.token-2.com" {"authentication" "standard"
                                           "cpu" 1
                                           "mem" 2048})
+    (kv/store kv-store "www.token-2s.com" {"authentication" "standard"
+                                          "cpu" 1
+                                          "mem" 2048
+                                           "profile" "service"})
```
[the remark this answers is not in the document]
Comment (shamsimam) [reply waiter/test/waiter/core_test.clj:1151]: No, will fix formatting issue.

File: waiter/test/waiter/core_test.clj [hunk 8]
```diff
@@ -1172,7 +1187,10 @@
       (let [test-request {:headers {"host" "www.token-1.com"}}
             {:keys [handled-request response]} (execute-request test-request)]
         (is (= (assoc test-request :waiter-discovery {:passthrough-headers {"host" "www.token-1.com"}
-                                                      :service-parameter-template {"mem" 2048}
+                                                      :service-description-template {"concurrency-level" 100
```
[the remark this answers is not in the document]
Comment (shamsimam) [reply waiter/test/waiter/core_test.clj:1190]: `cpu` isn't valid parameter, it was not intentional, will fix.

File: waiter/test/waiter/core_test.clj [hunk 9]
```diff
@@ -1191,19 +1213,92 @@
                                                                                sd/user-metadata-keys))
                                                       :waiter-headers {}})
                handled-request))
+        (is (= handler-response response)))
+
+      (let [test-request {:headers {"host" "www.token-2s.com"}}
+            {:keys [handled-request response]} (execute-request test-request)]
+        (is (= (assoc test-request :waiter-discovery {:passthrough-headers {"host" "www.token-2s.com"}
+                                                      :service-description-template {"authentication" "standard"
```
[the remark this answers is not in the document]
Comment (shamsimam) [reply waiter/test/waiter/core_test.clj:1223]: I had the context, the `p` or `s` in the name telling me which profile was activated, so I know what diff to expect in the results from the corresponding non-profile case.

File: waiter/test/waiter/core_test.clj (hunk 8 above)
[the remark this answers is not in the document]
Comment (shamsimam) [reply waiter/test/waiter/core_test.clj:1190]: Fixed.

File: waiter/test/waiter/core_test.clj (hunk 7 above)
[the remark this answers is not in the document]
Comment (shamsimam) [reply waiter/test/waiter/core_test.clj:1151]: Fixed.

File: waiter/src/waiter/service_description.clj (hunk 5 above)
[the remark this answers is not in the document]
Comment (sradack) [reply waiter/src/waiter/service_description.clj:1339]: Why do we bother with `discover-service-parameters` when we could use the `request->descriptor`?  I thought the answer was it was only for token-only workflows.

[the remark this answers is not in the document]
Comment (sradack) [reply waiter/src/waiter/service_description.clj:1339]: Can you point out any places where we use the values from headers to cause some change in behavior"?  Is it only to enable JWT on the fly?

[the remark this answers is not in the document]
Comment (shamsimam) [reply waiter/src/waiter/service_description.clj:1339]: > Why do we bother with discover-service-parameters when we could use the request->descriptor? I thought the answer was it was only for token-only workflows.

Service discovery is an early attempt at making the descriptor available before we have authenticated the user, the authenticated user is used to create the final descriptor.

> Can you point out some use cases that this change facilitates where we use the values from headers to cause some change in behavior? Is it only to enable JWT on the fly?

Anywhere affected by the `service-description-template` rename in this PR.
Two places, JWT and determining the auth method (standard vs saml).
The third affected place doesn't allow on-the-fly headers.

Screen Shot 2020-04-24 at 9 45 58 AM

Screen Shot 2020-04-24 at 9 48 26 AM
